catroid/src/main/java/org/catrobat/catroid/io/asynctask/ProjectSaver.kt

```diff
@@ -0,0 +1,100 @@
+/*
+ * Catroid: An on-device visual programming system for Android devices
+ * Copyright (C) 2010-2021 The Catrobat Team
+ * (<http://developer.catrobat.org/credits>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * An additional term exception under section 7 of the GNU Affero
+ * General Public License, version 3, is available at
+ * http://developer.catrobat.org/license_additional_term
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.catrobat.catroid.io.asynctask
+
+import android.content.Context
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import org.catrobat.catroid.content.LegoEV3Setting
+import org.catrobat.catroid.content.LegoNXTSetting
+import org.catrobat.catroid.content.Project
+import org.catrobat.catroid.content.bricks.Brick
+import org.catrobat.catroid.io.XstreamSerializer
+import org.catrobat.catroid.ui.settingsfragments.SettingsFragment
+import java.lang.ref.WeakReference
+
+class ProjectSaver(private val project: Project, context: Context) {
+    private val weakContextReference: WeakReference<Context> = WeakReference(context)
+    private var onProjectSaveListener: ProjectSaveListener? = null
+    private val scopeIO = CoroutineScope(Dispatchers.IO)
+
+    fun setOnProjectSaveListener(listener: ProjectSaveListener?) {
+        onProjectSaveListener = listener
+    }
+
+    fun saveProject() {
```

Comment:
```suggestion
    fun saveProjectAsync(listener: ProjectSaveListener?) {
```
Please rename. Dependency injection of the listener is not necessary, but please do it if possible. (It also reduces the probability of bugs and memory leaks (If an activity is passed or some other context)
If you want you can get rid of the ProjectSaveListener with "onSaveProjectComplete: (Boolean) -> Unit"

It's also possible to dependency inject the scope:
```suggestion
    fun saveProject(scope: CoroutineScope, onSaveProjectComplete: (Boolean)->Unit) {
```
in the tests, you can then use runBlocking.